sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/statsEstimation/JoinEstimationSuite.scala

```diff
@@ -67,6 +68,205 @@ class JoinEstimationSuite extends StatsEstimationTestBase {
+      HistogramBin(lo = 10, hi = 30, ndv = 10), HistogramBin(lo = 30, hi = 60, ndv = 30)))
+    val histogram2 = Histogram(height = 100, Array(
+      HistogramBin(lo = 0, hi = 50, ndv = 50), HistogramBin(lo = 50, hi = 100, ndv = 40)))
+    // test bin trimming
+    val (t1, h1) = trimBin(histogram2.bins(0), height = 100, min = 10, max = 60)
+    assert(t1 == HistogramBin(lo = 10, hi = 50, ndv = 40) && h1 == 80)
+    val (t2, h2) = trimBin(histogram2.bins(1), height = 100, min = 10, max = 60)
+    assert(t2 == HistogramBin(lo = 50, hi = 60, ndv = 8) && h2 == 20)
+
+    val expectedRanges = Seq(
+      OverlappedRange(10, 30, math.min(10, 40*1/2), math.max(10, 40*1/2), 300, 80*1/2),
+      OverlappedRange(30, 50, math.min(30*2/3, 40*1/2), math.max(30*2/3, 40*1/2), 300*2/3, 80*1/2),
+      OverlappedRange(50, 60, math.min(30*1/3, 8), math.max(30*1/3, 8), 300*1/3, 20)
+    )
+    assert(expectedRanges.equals(
+      getOverlappedRanges(histogram1, histogram2, newMin = 10D, newMax = 60D)))
+
+    estimateByHistogram(
+      histogram1 = histogram1,
+      histogram2 = histogram2,
+      expectedMin = 10D,
+      expectedMax = 60D,
+      // 10 + 20 + 8
+      expectedNdv = 38L,
+      // 300*40/20 + 200*40/20 + 100*20/10
+      expectedRows = 1200L)
+  }
+
+  test("equi-height histograms: a bin has only one value") {
+    val histogram1 = Histogram(height = 300, Array(
+      HistogramBin(lo = 30, hi = 30, ndv = 1), HistogramBin(lo = 30, hi = 60, ndv = 30)))
+    val histogram2 = Histogram(height = 100, Array(
+      HistogramBin(lo = 0, hi = 50, ndv = 50), HistogramBin(lo = 50, hi = 100, ndv = 40)))
+    // test bin trimming
+    val (t1, h1) = trimBin(histogram2.bins(0), height = 100, min = 30, max = 60)
+    assert(t1 == HistogramBin(lo = 30, hi = 50, ndv = 20) && h1 == 40)
+    val (t2, h2) = trimBin(histogram2.bins(1), height = 100, min = 30, max = 60)
+    assert(t2 ==HistogramBin(lo = 50, hi = 60, ndv = 8) && h2 == 20)
+
+    val expectedRanges = Seq(
+      OverlappedRange(30, 30, 1, 1, 300, 40/20),
+      OverlappedRange(30, 50, math.min(30*2/3, 20), math.max(30*2/3, 20), 300*2/3, 40),
+      OverlappedRange(50, 60, math.min(30*1/3, 8), math.max(30*1/3, 8), 300*1/3, 20)
+    )
+    assert(expectedRanges.equals(
+      getOverlappedRanges(histogram1, histogram2, newMin = 30D, newMax = 60D)))
+
+    estimateByHistogram(
+      histogram1 = histogram1,
+      histogram2 = histogram2,
+      expectedMin = 30D,
+      expectedMax = 60D,
+      // 1 + 20 + 8
+      expectedNdv = 29L,
+      // 300*20/1 + 200*40/20 + 100*20/10
+      expectedRows = 1200L)
+  }
+
+  test("equi-height histograms: a bin has only one value after trimming") {
+    val histogram1 = Histogram(height = 300, Array(
+      HistogramBin(lo = 50, hi = 60, ndv = 10), HistogramBin(lo = 60, hi = 75, ndv = 3)))
+    val histogram2 = Histogram(height = 100, Array(
+      HistogramBin(lo = 0, hi = 50, ndv = 50), HistogramBin(lo = 50, hi = 100, ndv = 40)))
```

Comment:
For the very skewed cases, multiple bins in a histogram may have same distinct value.  We may add one more test case to cover this situation.

Reply:
OK, I've added test cases for joins of skewed histograms (same skewed value and different skewed values).